File: backend/src/main/java/com/anushibinj/mypromptlib/service/PromptService.java
package com.anushibinj.mypromptlib.service;

import com.anushibinj.mypromptlib.model.Prompt;
import com.anushibinj.mypromptlib.model.PromptVersion;
import com.anushibinj.mypromptlib.repository.PromptRepository;
import com.anushibinj.mypromptlib.repository.PromptVersionRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class PromptService {

    private final PromptRepository promptRepository;
    private final PromptVersionRepository promptVersionRepository;

    public List<Prompt> getAllPromptsForUser(UUID userId) {
        return promptRepository.findByUserId(userId);
    }

    public List<Prompt> getPublicPrompts() {
        return promptRepository.findByIsPublicTrue();
    }

    public Prompt getPromptByIdAndUser(UUID id, UUID userId) {
        Prompt prompt = promptRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Prompt not found"));
        if (!prompt.getUserId().equals(userId)) {
            throw new RuntimeException("Unauthorized to access this prompt");
        }
        return prompt;
    }

    public Prompt getSharedPrompt(UUID id) {
        Prompt prompt = promptRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Prompt not found"));
        if (!prompt.isPublic()) {
            throw new RuntimeException("This prompt is not public");
        }
        return prompt;
    }

    @Transactional
    public Prompt createPrompt(Prompt prompt, UUID userId) {
        prompt.setUserId(userId);
        Prompt saved = promptRepository.save(prompt);
        saveVersion(saved);
        return saved;
    }

    @Transactional
    public Prompt updatePrompt(UUID id, Prompt promptDetails, UUID userId) {
        Prompt existing = getPromptByIdAndUser(id, userId);
        existing.setTitle(promptDetails.getTitle());
        existing.setContent(promptDetails.getContent());
        existing.setPublic(promptDetails.isPublic());
        Prompt saved = promptRepository.save(existing);
        saveVersion(saved);
        return saved;
    }

    @Transactional
    public void deletePrompt(UUID id, UUID userId) {
        Prompt existing = getPromptByIdAndUser(id, userId);
        promptVersionRepository.deleteByPromptId(id);
        promptRepository.delete(existing);
    }

    public List<PromptVersion> getPromptHistory(UUID promptId, UUID userId) {
        // Verify ownership
        getPromptByIdAndUser(promptId, userId);
        return promptVersionRepository.findByPromptIdOrderByVersionNumberDesc(promptId);
    }

    private void saveVersion(Prompt prompt) {
        int nextVersion = promptVersionRepository
                .findTopByPromptIdOrderByVersionNumberDesc(prompt.getId())
                .map(v -> v.getVersionNumber() + 1)
                .orElse(1);

        PromptVersion version = PromptVersion.builder()
                .promptId(prompt.getId())
                .versionNumber(nextVersion)
                .title(prompt.getTitle())
                .content(prompt.getContent())
                .isPublic(prompt.isPublic())
                .createdAt(Instant.now())
                .build();
        promptVersionRepository.save(version);
    }
}

File: backend-go/internal/service/prompt_service.go
package service

import (
	"errors"
	"time"

	"backend-go/internal/model"
	"backend-go/internal/repository"

	"github.com/google/uuid"
)

type PromptService struct {
	promptRepo        *repository.PromptRepository
	promptVersionRepo *repository.PromptVersionRepository
}

func NewPromptService(promptRepo *repository.PromptRepository, promptVersionRepo *repository.PromptVersionRepository) *PromptService {
	return &PromptService{
		promptRepo:        promptRepo,
		promptVersionRepo: promptVersionRepo,
	}
}

func (s *PromptService) GetAllPromptsForUser(userId uuid.UUID) ([]model.Prompt, error) {
	return s.promptRepo.FindByUserID(userId)
}

func (s *PromptService) GetPublicPrompts() ([]model.Prompt, error) {
	return s.promptRepo.FindByIsPublicTrue()
}

func (s *PromptService) GetPromptByIdAndUser(id uuid.UUID, userId uuid.UUID) (*model.Prompt, error) {
	prompt, err := s.promptRepo.FindByID(id)
	if err != nil {
		return nil, err
	}
	if prompt == nil {
		return nil, errors.New("Prompt not found")
	}
	if prompt.UserID != userId {
		return nil, errors.New("Unauthorized to access this prompt")
	}
	return prompt, nil
}

func (s *PromptService) GetSharedPrompt(id uuid.UUID) (*model.Prompt, error) {
	prompt, err := s.promptRepo.FindByID(id)
	if err != nil {
		return nil, err
	}
	if prompt == nil {
		return nil, errors.New("Prompt not found")
	}
	if !prompt.IsPublic {
		return nil, errors.New("This prompt is not public")
	}
	return prompt, nil
}

func (s *PromptService) CreatePrompt(prompt *model.Prompt, userId uuid.UUID) (*model.Prompt, error) {
	// ID generation handles by default or model
	if prompt.ID == uuid.Nil {
		prompt.ID = uuid.New()
	}
	prompt.UserID = userId
	saved, err := s.promptRepo.Save(prompt)
	if err != nil {
		return nil, err
	}
	err = s.saveVersion(saved)
	if err != nil {
		return nil, err
	}
	return saved, nil
}

func (s *PromptService) UpdatePrompt(id uuid.UUID, promptDetails *model.Prompt, userId uuid.UUID) (*model.Prompt, error) {
	existing, err := s.GetPromptByIdAndUser(id, userId)
	if err != nil {
		return nil, err
	}

	existing.Title = promptDetails.Title
	existing.Content = promptDetails.Content
	existing.IsPublic = promptDetails.IsPublic

	saved, err := s.promptRepo.Save(existing)
	if err != nil {
		return nil, err
	}

	err = s.saveVersion(saved)
	if err != nil {
		return nil, err
	}

	return saved, nil
}

func (s *PromptService) DeletePrompt(id uuid.UUID, userId uuid.UUID) error {
	existing, err := s.GetPromptByIdAndUser(id, userId)
	if err != nil {
		return err
	}

	err = s.promptVersionRepo.DeleteByPromptID(id)
	if err != nil {
		return err
	}

	return s.promptRepo.Delete(existing)
}

func (s *PromptService) GetPromptHistory(promptId uuid.UUID, userId uuid.UUID) ([]model.PromptVersion, error) {
	// Verify ownership
	_, err := s.GetPromptByIdAndUser(promptId, userId)
	if err != nil {
		return nil, err
	}

	return s.promptVersionRepo.FindByPromptIDOrderByVersionNumberDesc(promptId)
}

func (s *PromptService) saveVersion(prompt *model.Prompt) error {
	nextVersion := 1
	topVersion, err := s.promptVersionRepo.FindTopByPromptIDOrderByVersionNumberDesc(prompt.ID)
	if err != nil {
		return err
	}
	if topVersion != nil {
		nextVersion = topVersion.VersionNumber + 1
	}

	version := &model.PromptVersion{
		ID:            uuid.New(),
		PromptID:      prompt.ID,
		VersionNumber: nextVersion,
		Title:         prompt.Title,
		Content:       prompt.Content,
		IsPublic:      prompt.IsPublic,
		CreatedAt:     time.Now(), // time.Now().UTC() matches Instant.now() depending on setup, but time.Now is standard
	}

	_, err = s.promptVersionRepo.Save(version)
	return err
}
